src/components/RangeDateField/hooks/useRangeDateFieldState.ts

```diff
@@ -0,0 +1,306 @@
+/* eslint-disable complexity */
+import React from 'react';
+
+import {isValid} from '@gravity-ui/date-utils';
+import type {DateTime} from '@gravity-ui/date-utils';
+
+import {
+    type BaseDateFieldState,
+    useBaseDateFieldState,
+} from '../../DateField/hooks/useBaseDateFieldState';
+import type {DateFieldSectionType, DateFieldSectionWithoutPosition} from '../../DateField/types';
+import {
+    EDITABLE_SEGMENTS,
+    addSegment,
+    parseDateFromString,
+    setSegment,
+    splitFormatIntoSections,
+} from '../../DateField/utils';
+import {useControlledState} from '../../hooks/useControlledState';
+import type {DateFieldBase} from '../../types/datePicker';
+import type {RangeValue} from '../../types/inputs';
+import {createPlaceholderValue, isInvalid} from '../../utils/dates';
+import {getRangeEditableSections} from '../utils';
+
+export interface RangeDateFieldStateOptions extends DateFieldBase<RangeValue<DateTime>> {
+    delimeter?: string;
+}
+
+const RANGE_FORMAT_DELIMETER = ' — ';
+
+export type RangeDateFieldState = BaseDateFieldState<RangeValue<DateTime>>;
+
+export function useRangeDateFieldState(props: RangeDateFieldStateOptions): RangeDateFieldState {
+    const [value, setDate] = useControlledState(props.value, props.defaultValue, props.onUpdate);
+
+    const [placeholderDate, setPlaceholderDate] = React.useState<RangeValue<DateTime>>(() => {
+        const date = createPlaceholderValue({
+            placeholderValue: props.placeholderValue,
+            timeZone: props.timeZone,
+        });
+
+        return {start: date, end: date};
+    });
+
+    const format = props.format || 'L';
+    const delimeter = props.delimeter ?? RANGE_FORMAT_DELIMETER;
+    const sections = useFormatSections(format);
+
+    const allSegments: typeof EDITABLE_SEGMENTS = React.useMemo(() => {
+        return sections
+            .filter((seg) => EDITABLE_SEGMENTS[seg.type])
+            .reduce<typeof EDITABLE_SEGMENTS>((p, seg) => ({...p, [seg.type]: true}), {});
+    }, [sections]);
+
+    // eslint-disable-next-line prefer-const
+    let [validSegments, setValidSegments] = React.useState<RangeValue<typeof EDITABLE_SEGMENTS>>(
```

Comment:
I think it is not a good idea to mutate `validSegments`

Reply:
Yes. But that's how it works in DateField, and I'm afraid of breaking the logic.

Reply:
But I'll check it.